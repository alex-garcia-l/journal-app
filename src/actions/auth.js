import Swal from 'sweetalert2';

import { types } from '../types/types';
import { finishLoading, startLoading } from './ui';
import { firebase, googleAuthProvider } from '../database/firebase-config';

export const startLogin = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    firebase.auth().signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch(error => {
        dispatch(finishLoading());
        Swal.fire('Error', error.message, 'error');
      });
  }
}

export const startLoginWithGoogle = () => {
  return (dispatch) => {
    firebase.auth().signInWithPopup(googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
      });
  }
}

export const startUserRegister = (name, email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    firebase.auth().createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch(error => {
        dispatch(finishLoading());
        Swal.fire('Error', error.message, 'error');
      });
  }
}

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout());
  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
})

export const logout = () => ({
  type: types.logout
})
