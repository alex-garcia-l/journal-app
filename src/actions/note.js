import Swal from 'sweetalert2';
import { db } from '../database/firebase-config';
import { types } from '../types/types';
import { loadNotes } from '../helpers/loadNotes';
import { uploadImage } from '../helpers/uploadImage';

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }

    const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
    dispatch(setNoteActive(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote));
  }
}

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  }
}

export const startUpdateNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    if (!note.url) {
      delete note.url;
    }

    const noteUpdate = { ...note };
    delete noteUpdate.id;

    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteUpdate);

    dispatch(refreshNotes(note.id, noteUpdate));

    Swal.fire({
      title: 'Success',
      text: 'Note modified',
      icon: 'success',
      showConfirmButton: false,
      timer: 1500
    });
  }
}

export const startUploadImage = (file) => {
  return async (dispatch, getState) => {
    const { active: note } = getState().notes;

    Swal.fire({
      title: 'Uploading',
      text: 'Please wait...',
      allowOutsideClick: false,
      allowEscapeKey: false,
      allowEnterKey: false,
      didOpen: () => {
        Swal.showLoading();
      },
      willClose: () => {
        Swal.hideLoading()
      },
    });

    const upload = await uploadImage(file);
    note.url = upload.secure_url;
    dispatch(startUpdateNote(note));

    Swal.close();
  }
}

export const startDeleteNote = (id) => {
  return (dispatch, getState) => {

    Swal.fire({
      title: 'Are you sure you want to delete the note?',
      confirmButtonText: 'Yes, delete',
      showCancelButton: true,
      icon: 'question',
    }).then(async (result) => {
      /* Read more about isConfirmed, isDenied below */
      if (result.isConfirmed) {
        const { uid } = getState().auth;

        await db.doc(`${uid}/journal/notes/${id}`).delete();
        dispatch(deleteNote(id));
      }
    })
  }
}

export const setNoteActive = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note
  }
});

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note
  }
});

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes
});

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id
});

export const refreshNotes = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: {
      id,
      ...note
    }
  }
});

export const purgeNotes = () => ({
  type: types.notesLogoutCleaning
})
