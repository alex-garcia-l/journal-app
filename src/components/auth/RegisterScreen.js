import React from 'react';
import { Link } from 'react-router-dom';

export const RegisterScreen = () => {
  return (
    <>
      <h3 className="auth__title mb-5">
        <i class="fa-solid fa-user-plus fa-3x mb-5"></i>
        <span>Register</span>
      </h3>
      
      <form>
        <input
          type="text"
          placeholder="Name"
          name="name"
          autoComplete="off"
          className="input"
        />

        <input
          type="email"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="input"
        />

        <input
          type="password"
          placeholder="Password"
          name="password"
          className="input"
        />

        <input
          type="password"
          placeholder="Confirm password"
          name="password_confirm"
          className="input"
        />

        <button
          type="submit"
          className="btn btn-primary btn-block"
        >
          Register
        </button>

        <div className="auth__create_acount">
          <Link
            to="/auth/login"
          >
            Already registered?
          </Link>
        </div>
      </form>
    </>
  );
};
