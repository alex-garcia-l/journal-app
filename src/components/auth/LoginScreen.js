import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLogin, startLoginWithGoogle } from '../../actions/auth';

import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {

  const dispatch = useDispatch();
  const { loading } = useSelector(state => state.ui);

  const [valuesForm, handleInputChange] = useForm({
    email: '',
    password: ''
  });

  const { email, password } = valuesForm;

  const handleSubmit = (evt) => {
    evt.preventDefault();
    dispatch(startLogin(email, password));
  }

  const handleGoogleLogin = () => {
    dispatch(startLoginWithGoogle());
  }

  return (
    <div className="animate__animated animate__fadeIn animate__faster">
      <h3 className="auth__title mb-5">
        <i className="fa-solid fa-user-large fa-3x mb-5"></i>
        <span>Login</span>
      </h3>
      <form onSubmit={handleSubmit}>
        <input
          type="email"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="input"
          value={email}
          onChange={handleInputChange}
        />

        <input
          type="password"
          placeholder="Password"
          name="password"
          className="input"
          value={password}
          onChange={handleInputChange}
        />
        <button
          type="submit"
          className="btn btn-primary btn-block"
          disabled={loading}
        >
          Login
        </button>

        <div className="auth__social_networks">
          <p className="auth__paragraph">
            Login with social network
          </p>
          <div className="google-btn" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <div className="auth__create_acount">
          <Link
            to="/auth/register"
          >
            Create account
          </Link>
        </div>
      </form>
    </div>
  );
};
