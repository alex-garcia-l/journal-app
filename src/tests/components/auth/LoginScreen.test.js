/**
 * @jest-environment node
 */

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLoginWithGoogle, startLogin } from '../../../actions/auth';

import 'jsdom-global/register';

jest.mock('../../../actions/auth', () => ({
  startLoginWithGoogle: jest.fn(),
  startLogin: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
  auth: {},
  ui: {
    loading: false,
    msgError: null
  }
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter >
      <LoginScreen />
    </MemoryRouter>
  </Provider >
);

describe('Pruebas para <LoginScreen />', () => {

  beforeEach(() => {
    store = mockStore(initialState);
    jest.clearAllMocks();
  });

  test('debe renderizar correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe disparar la acciÃ³n de handleGoogleLogin', () => {

    wrapper.find('.google-btn').prop('onClick')();

    expect(startLoginWithGoogle).toHaveBeenCalled();
  });

  test('debe disparar el startLogin con sus respectivos valores', () => {

    const email = 'mail@mail.com';
    const password = '1234abcd';

    wrapper.find({ name: "email" }).simulate('change', {
      target: {
        name: 'email',
        value: email
      }
    });

    wrapper.find({ name: "password" }).simulate('change', {
      target: {
        name: 'password',
        value: password
      }
    });

    wrapper.find('form').prop('onSubmit')({ preventDefault: () => { } });

    expect(startLogin).toHaveBeenCalledWith(email, password);
  })

});
