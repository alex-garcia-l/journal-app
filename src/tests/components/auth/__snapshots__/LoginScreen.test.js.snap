// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas para <LoginScreen /> debe renderizar correctamente 1`] = `
<div
  className="animate__animated animate__fadeIn animate__faster"
>
  <h3
    className="auth__title mb-5"
  >
    <i
      className="fa-solid fa-user-large fa-3x mb-5"
    />
    <span>
      Login
    </span>
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value=""
    />
    <input
      className="input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block"
      disabled={false}
      type="submit"
    >
      Login
    </button>
    <div
      className="auth__social_networks"
    >
      <p
        className="auth__paragraph"
      >
        Login with social network
      </p>
      <div
        className="google-btn"
        onClick={[Function]}
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Sign in with google
          </b>
        </p>
      </div>
    </div>
    <div
      className="auth__create_acount"
    >
      <a
        href="/auth/register"
        onClick={[Function]}
      >
        Create account
      </a>
    </div>
  </form>
</div>
`;
